metadata-ingestion/src/datahub/ingestion/source/redshift/lineage.py

```diff
@@ -613,3 +649,108 @@ def get_lineage(
     def report_status(self, step: str, status: bool) -> None:
         if self.redundant_run_skip_handler:
             self.redundant_run_skip_handler.report_current_run_status(step, status)
+
+    @lru_cache(maxsize=64)
+    def get_temp_tables(
+        self, connection: redshift_connector.Connection
+    ) -> List[TempTableRow]:
+        ddl_query: str = RedshiftQuery.temp_table_ddl_query(
+            start_time=self.config.start_time,
+            end_time=self.config.end_time,
+        )
+
+        logger.debug(f"Temporary table ddl query = {ddl_query}")
+
+        temp_table_rows: List[TempTableRow] = [
+            row
+            for row in RedshiftDataDictionary.get_temporary_rows(
+                conn=connection,
+                query=ddl_query,
+            )
+        ]
+
+        logger.debug(f"Number of temp tables = {len(temp_table_rows)}")
+
+        return temp_table_rows
+
+    def find_temp_tables(
+        self, connection: redshift_connector.Connection, temp_table_names: List[str]
+    ) -> List[TempTableRow]:
+
+        matched_temp_tables: List[TempTableRow] = []
+
+        for table_name in temp_table_names:
+            for row in self.get_temp_tables(connection=connection):
+                if any(
+                    row.query_text.lower().startswith(prefix)
+                    for prefix in RedshiftQuery.get_temp_table_clause(table_name)
+                ):
+                    matched_temp_tables.append(row)
+
+        return matched_temp_tables
+
+    def _add_permanent_datasets_recursively(
+        self,
+        db_name: str,
+        temp_table_rows: List[TempTableRow],
+        connection: redshift_connector.Connection,
+        permanent_lineage_datasets: List[LineageDataset],
+    ) -> None:
+
+        transitive_temp_tables: List[TempTableRow] = []
+
+        for temp_table in temp_table_rows:
+            intermediate_l_datasets, _ = self._get_sources_from_query(
+                db_name=db_name,
+                query=temp_table.query_text,
+            )
+
+            # make sure lineage dataset should not contain a temp table
+            # if such dataset is present then add it to transitive_temp_tables to resolve it to original permanent table
+            for lineage_dataset in intermediate_l_datasets:
+
+                db, schema, table = split_qualified_table_name(lineage_dataset.urn)
+
+                # Check if table found is again a temp table
+                repeated_temp_table: List[TempTableRow] = self.find_temp_tables(
+                    connection=connection,
+                    temp_table_names=[table],
+                )
+
+                if repeated_temp_table:
+                    transitive_temp_tables.extend(repeated_temp_table)
```

Comment:
make sure you don't add the same table name twice (for example if you have a self-join) otherwise it will kill your recursion loop